cli/loader/src/lib.rs

```diff
@@ -93,11 +93,26 @@ pub struct LanguageConfiguration<'a> {
     use_all_highlight_names: bool,
 }
 
+#[derive(Debug)]
+struct Regex2(Regex);
+impl Eq for Regex2 {}
+impl PartialEq for Regex2 {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.as_str() == other.0.as_str()
+    }
+}
+impl std::hash::Hash for Regex2 {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.0.as_str().hash(state)
+    }
+}
+
```

Comment:
It'd be better to do away with this and perhaps just use the string version of the regex for the hashmap entry, and then use Regex::from_str to convert back when using `find`

Reply:
Here's what I was thinking:

```patch
diff --git a/cli/loader/src/lib.rs b/cli/loader/src/lib.rs
index 0f9a33fe..88c47986 100644
--- a/cli/loader/src/lib.rs
+++ b/cli/loader/src/lib.rs
@@ -8,6 +8,7 @@ use std::io::BufReader;
 use std::ops::Range;
 use std::path::{Path, PathBuf};
 use std::process::Command;
+use std::str::FromStr;
 use std::sync::Mutex;
 use std::time::SystemTime;
 use std::{env, fs, mem};
@@ -93,26 +94,12 @@ pub struct LanguageConfiguration<'a> {
     use_all_highlight_names: bool,
 }
 
-#[derive(Debug)]
-struct Regex2(Regex);
-impl Eq for Regex2 {}
-impl PartialEq for Regex2 {
-    fn eq(&self, other: &Self) -> bool {
-        self.0.as_str() == other.0.as_str()
-    }
-}
-impl std::hash::Hash for Regex2 {
-    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
-        self.0.as_str().hash(state)
-    }
-}
-
 pub struct Loader {
     parser_lib_path: PathBuf,
     languages_by_id: Vec<(PathBuf, OnceCell<Language>)>,
     language_configurations: Vec<LanguageConfiguration<'static>>,
     language_configuration_ids_by_file_type: HashMap<String, Vec<usize>>,
-    language_configuration_ids_by_first_line_regex: HashMap<Regex2, Vec<usize>>,
+    language_configuration_ids_by_first_line_regex: HashMap<String, Vec<usize>>,
     highlight_names: Box<Mutex<Vec<String>>>,
     use_all_highlight_names: bool,
     debug_build: bool,
@@ -238,14 +225,22 @@ impl Loader {
                     })
             })
             .or_else(|| {
-                let Ok(file) = fs::File::open(path) else {return None};
+                let Ok(file) = fs::File::open(path) else {
+                    return None;
+                };
                 let reader = BufReader::new(file);
-                let Some(Ok(first_line)) = std::io::BufRead::lines(reader).next() else {return None};
+                let Some(Ok(first_line)) = std::io::BufRead::lines(reader).next() else {
+                    return None;
+                };
 
                 self.language_configuration_ids_by_first_line_regex
                     .iter()
-                    .find(|(regex,_)| dbg!(&regex.0).is_match(&first_line))
-                    .map(|(_,ids)| ids)
+                    .find(|(regex, _)| {
+                        Regex::from_str(regex)
+                            .and_then(|r| Ok(r.is_match(&first_line)))
+                            .unwrap_or(false)
+                    })
+                    .map(|(_, ids)| ids)
             });
 
         if let Some(configuration_ids) = configuration_ids {
@@ -616,7 +611,7 @@ impl Loader {
                     }
                     for first_line_regex in &configuration.first_line_regex {
                         self.language_configuration_ids_by_first_line_regex
-                            .entry(Regex2(first_line_regex.clone()))
+                            .entry(first_line_regex.to_string())
                             .or_insert(Vec::new())
                             .push(self.language_configurations.len());
                     }

```

Reply:
Should I just change https://github.com/tree-sitter/tree-sitter/blob/e854fc4731c20d56a6212569c36823a1aa84b166/cli/loader/src/lib.rs#L81 to String

Instead of the current Regex -> String -> Regex ?

Reply:
we loose the regex validation though

Reply:
Up to you, but I don't really like the whole abstraction of a struct on top of regex..